crates/bitcoin-da/src/helpers/builders.rs

```diff
@@ -457,38 +957,235 @@ pub fn create_inscription_transactions(
+                    id: reveal_tx.compute_txid(),
+                    tx: reveal_tx,
+                },
+            });
+        }
+
+        nonce += 1;
+    }
+}
 
+// TODO: parametrize hardness
+// so tests are easier
+// Creates the batch proof transactions Type 0 - BatchProvingTxs - SequencerCommitment
+#[allow(clippy::too_many_arguments)]
+#[instrument(level = "trace", skip_all, err)]
+pub fn create_batchproof_type_0(
+    rollup_name: &[u8],
+    body: Vec<u8>,
+    da_private_key: &SecretKey,
+    prev_utxo: Option<UTXO>,
+    utxos: Vec<UTXO>,
+    recipient: Address,
+    reveal_value: u64,
+    commit_fee_rate: f64,
+    reveal_fee_rate: f64,
+    network: Network,
+    reveal_tx_prefix: &[u8],
+) -> Result<BatchProvingTxs, anyhow::Error> {
+    debug_assert!(
+        body.len() < 520,
+        "The body of a serialized sequencer commitment exceeds 520 bytes"
+    );
+    // Create commit key
+    let secp256k1 = Secp256k1::new();
+    let key_pair = UntweakedKeypair::new(&secp256k1, &mut rand::thread_rng());
+    let (public_key, _parity) = XOnlyPublicKey::from_keypair(&key_pair);
+
+    let header = TransactionHeaderBatchProof {
+        rollup_name,
+        kind: TransactionKindBatchProof::SequencerCommitment,
+    };
+    let header_bytes = header.to_bytes();
+
+    // sign the body for authentication of the sequencer
+    let (signature, signer_public_key) =
+        sign_blob_with_private_key(&body, da_private_key).expect("Sequencer sign the body");
+
+    // start creating inscription content
+    let reveal_script_builder = script::Builder::new()
+        .push_x_only_key(&public_key)
+        .push_opcode(OP_CHECKSIGVERIFY)
+        .push_slice(PushBytesBuf::try_from(header_bytes).expect("Cannot push header"))
+        .push_opcode(OP_FALSE)
+        .push_opcode(OP_IF)
+        .push_slice(PushBytesBuf::try_from(signature).expect("Cannot push signature"))
+        .push_slice(
+            PushBytesBuf::try_from(signer_public_key).expect("Cannot push sequencer public key"),
+        )
+        .push_slice(PushBytesBuf::try_from(body).expect("Cannot push sequencer commitment"))
+        .push_opcode(OP_ENDIF);
+
+    // Start loop to find a 'nonce' i.e. random number that makes the reveal tx hash starting with zeros given length
+    let mut nonce: i64 = 16; // skip the first digits to avoid OP_PUSHNUM_X
+    loop {
+        if nonce % 10000 == 0 {
+            trace!(nonce, "Trying to find commit & reveal nonce");
+            if nonce > 65536 {
+                warn!("Too many iterations finding nonce");
+            }
+        }
+        let utxos = utxos.clone();
+        let recipient = recipient.clone();
+        // ownerships are moved to the loop
+        let mut reveal_script_builder = reveal_script_builder.clone();
+
+        // push first random number and body tag
+        reveal_script_builder = reveal_script_builder
+            .push_slice(nonce.to_le_bytes())
+            .push_opcode(OP_DROP);
```

Comment:
just now realized that we can actually have the nonce inside if / endif for sequencer commitments since we know the body is a single push so we don't need to do OP_DROP.

if this is correct please open an issue we can merge like this np

Reply:
It is true but parsing in my case looks nice and neat. In case of nonce in the body we need to take pushbuf except for the last one and the last one will be nonce

Reply:
hmm, resolve this conversation but pls first open issue with low priority, we probably won't implement it right away. but maybe one day.

Reply:
https://github.com/chainwayxyz/citrea/issues/1024